rpc/types/README.md

```diff
@@ -0,0 +1,60 @@
+Monero RPC types.
+
+# What
+This crate ports the types used in Monero's RPC interface, including:
+- JSON types
+- Binary (epee) types
+- Mixed types
+- Other commonly used RPC types
+
+# Modules
+This crate's types are split in the following manner:
+
+This crate has 4 modules:
+- The root module; `cuprate_rpc_types`
+- [`json`] module; JSON types from the `/json_rpc` endpoint
+- [`bin`] module; Binary types from the binary endpoints
+- [`other`] module; Misc JSON types from other endpoints
+
+Miscellaneous types are found in the root module, e.g. [`crate::Status`].
+
+Each type in `{json,bin,other}` come in pairs and have identical names, but are suffixed with either `Request` or `Response`. e.g. [`GetBlockCountRequest`](crate::json::GetBlockCountRequest) & [`GetBlockCountResponse`](crate::json::GetBlockCountResponse).
+
+# Documentation
+The documentation for types within `{json,bin,other}` are omitted, as they can be found in [Monero's RPC documentation](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html).
+
+However, each type will document:
+- **Definition**: the exact type definition location in `monerod`
+- **Documentation**: the Monero RPC documentation link
+- **Request/response**: the other side of this type, either the request or response
+
+# Naming
+The naming for types within `{json,bin,other}` follow the following scheme:
+- Convert the endpoint or method name into `UpperCamelCase`
+- Remove any suffix extension
+
+For example:
+
+| Endpoint/method | Crate location and name |
+|-----------------|-------------------------|
+| [`get_block_count`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_block_count) | [`json::GetBlockCountRequest`] & [`json::GetBlockCountResponse`]
+| [`/get_blocks.bin`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_blockbin) | `bin::GetBlocksRequest` & `bin::GetBlocksResponse`
+| [`/get_height`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_height) | `other::GetHeightRequest` & `other::GetHeightResponse`
+
+TODO: fix doc links when types are ready.
+
+# Mixed types
+Note that some types within [`other`] mix JSON & binary together, i.e.,
+the message overall is JSON, however some fields contain binary
+values, for example:
+
+```json
+{
+  "string": "",
+  "float": 30.0,
+  "integer": 30,
+  "binary": /* serialized binary */
```

Comment:
Is it decided on how we are going to handle this yet?

Reply:
I know I mentioned using a JSON5 serde lib but I don't really want to have to pull in a lib like that in for this.

Reply:
AFAICT `monerod` puts binary inside JSON strings, e.g. this is the response from `get_txpool_backlog`:
```json
{
  "id": 0,
  "jsonrpc": "2.0",
  "result": {
    "backlog": "�\b����������", // note the quotes, it's a JSON string
    "credits": 0,
    "status": "OK",
    "top_hash": "",
    "untrusted": false
  }
}
```
So I think we just model it as a `String` Or maybe some custom wrapper around `Vec<u8>` so we don't pay UTF-8 costs? The following code works:
```rust
use serde::*;
use serde_json::*;

const JSON: &str = r#"{
  "id": 0,
  "jsonrpc": "2.0",
  "result": {
    "backlog": "�\b����������",
    "credits": 0,
    "status": "OK",
    "top_hash": "",
    "untrusted": false
  }
}"#;

#[derive(Serialize, Deserialize)]
struct GetTxpoolBacklog {
    id: u64,
    jsonrpc: String,
    result: Result,
}

#[derive(Serialize, Deserialize)]
struct Result {
    backlog: String, // a JSON string holding binary data
    status: String,
    untrusted: bool,
}

#[derive(EpeeStuff)]
struct Backlog {/*...*/}

fn main() {
    let s = from_str::<GetTxpoolBacklog>(JSON).unwrap(); // this works

    let epee = epee_decode::<Backlog>(&s.result.backlog.as_bytes());
}
```

Reply:
@SyntheticBird45 mentioned that there were some methods that return the raw binary in a field, not in a string.

I just had a look and it seems that the ones they mentioned do return it in a string unless I am missing something @SyntheticBird45?

Reply:
https://github.com/Cuprate/cuprate/pull/147#discussion_r1653774069